src/ILVerify/src/ILImporter.Verify.cs

```diff
@@ -635,6 +772,87 @@ void CheckIsComparable(StackValue a, StackValue b, ILOpcode op)
             }
         }
 
+        void CheckDelegateCreation(StackValue ftn, StackValue obj)
+        {
+            if (!_delegateCreateStart.HasValue)
+            {
+                VerificationError(VerifierError.DelegatePattern);
+                return;
+            }
+
+            int delegateStart = _delegateCreateStart.Value;
+
+            if (_currentInstructionOffset - delegateStart == 6) // ldftn <tok> takes 6 bytes
+            {
+                if (GetOpcodeAt(delegateStart) != ILOpcode.ldftn)
+                {
+                    VerificationError(VerifierError.DelegatePattern);
+                    return;
+                }
+                else
+                {
+                    // See "Rules for non-virtual call to a non-final virtual method" in ImportCall
+                    var owningTypeDef = (EcmaType)ftn.Method.OwningType.GetTypeDefinition();
```

Comment:
Might be better to cast to `MetadataType` here, just in case we would run this on a type system backed by a different metadata source (e.g. to verify code and synthetic helper types generated by the CoreRT ahead of time compiler).

Reply:
True, thanks.
I am currently also working on implementing the method / type access checks, where I need the Type-/MethodAttributes in order to get the necessary access information. Is there a better way to retrieve this information other than casting to `EcmaType` / `EcmaMethod`?